### PR TITLE
Feature: multiple raid fleets with more flexibility, and accounting for local attraction

### PR DESCRIPTION
**Pull request by Hurleveur**
**Feature:** This PR implements part of the feature request detailed and discussed in issue #7894, and also addresses a big early game issue (that I saw in a steam review)

## Feature Details
This allows definition of multiple raid fleets per government, each with their own minimal attraction (it defaults to the old 2 min attraction) as well as maximal attractions (if you are above that they wont spawn).

On top of that, it considers local fleets in the calculation for attraction, meaning that:
-you are more likely to get attacked by pirate raids in pirate systems (duh)
-you are less likely to get attacked by pirates in well defended systems
Pirates are getting a brain of their own, ik, crazy.

Other than that nothing changes with the raid fleets themselves, that would be for another PR. (for instance to make strength be related to deterrence of ships).

This also adds the `"raid chance in system: <system>"` autocondition, that is the chance of getting attacked by one fleet. It is multiplied by 1000 for extra precision.

## UI Screenshots
n/a

## Usage Examples
This means that we can make small, medium, big pirate raiding fleets that would attack at the relevant times (and combine with each others, making raids less of a joke.)
```html
raid <Fleet> <minimal attraction> <maximal attraction>
```
```html
government "Free World"
  raid "Small Southern Pirates"
  raid "Large Southern Pirates" 5 (minimum attraction needed, default to 2)
```
Also could prevent a noob with a small fleet from getting attacked by the same fleets as smbd with a huge fleet of ships.
We can also have multiple governments raiding one system.

The old man could just shout at you if you have a big chance to get hit by a fleet instead of just if you are safe in the Earth system with high cargo.

## Testing Done
I tested this long ago and it was working, should still work since I simplified the PR too.

### Automated Tests Added
n/a

## Performance Impact
n/a

### CODE REVIEW

**Comment by bene-dictator** [comment]
If this gets merged, is the pirate raid fleet necessary and would we just use the other pirate fleets?

**Comment by Hurleveur** [comment]
We could just use other fleets indeed. We could also stack one fleet multiple times to increase the odds of it showing up too, if we want to, the possibilities are endless.

File: source/Engine.cpp [hunk 1]
```diff
@@ -1261,20 +1261,28 @@ void Engine::EnterSystem()
...
-		double attraction = .005 * (factors.first - factors.second - 2.);
-		if(attraction > 0.)
-			for(int i = 0; i < 10; ++i)
-				if(Random::Real() < attraction)
-				{
-					raidFleet->Place(*system, newShips);
-					Messages::Add("Your fleet has attracted the interest of a "
-							+ raidGovernment->GetName() + " raiding party.", Messages::Importance::Highest);
-				}
+	const vector<pair<const Fleet *, double>> raidFleets = system->GetGovernment()->RaidFleets();
```
Comment (quyykk) [inline source/Engine.cpp:1264]: you don't need this variable

File: source/Engine.cpp [hunk 2]
```diff
@@ -1261,20 +1261,28 @@ void Engine::EnterSystem()
...
+		const Government *raidGovernment = raidFleet.first->GetGovernment();
+		if(raidGovernment && raidGovernment->IsEnemy())
+		{
+			pair<double, double> factors = player.RaidFleetFactors();
+			double attraction = .005 * (factors.first - factors.second - raidFleet.second);
+			// Consider the other fleets in the local system.
+			// Do not take into account the raid fleet, that would just mess with the minimum attraction of it.
+			for(const auto &fleet : system->Fleets())
+				attraction -= sqrt(fleet.Get()->Strength() / 10000.) / fleet.Period()
+					* (fleet.Get()->GetGovernment()->IsEnemy(raidGovernment) ? 1. : -1.);
```
Comment (quyykk) [inline source/Engine.cpp:1276]: Why would friendly fleets increase attraction?

File: source/Government.cpp [hunk 3]
```diff
@@ -258,7 +258,17 @@ void Government::Load(const DataNode &node)
 		else if(key == "language")
 			language = child.Token(valueIndex);
 		else if(key == "raid")
-			raidFleet = GameData::Fleets().Get(child.Token(valueIndex));
+		{
+			if(!add)
+				raidFleets.clear();
```
Comment (quyykk) [inline source/Government.cpp:263]: If I specify two raids in a single system and the second one shouldn't overwrite the first one.

File: source/Panel.cpp [hunk 4]
```diff
@@ -22,8 +22,8 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 #include "GameData.h"
 #include "Point.h"
 #include "Preferences.h"
-#include "Screen.h"
 #include "text/Format.h"
+#include "Screen.h"
```
Comment (quyykk) [inline source/Panel.cpp:26]: unrelated change

File: source/Engine.cpp (hunk 2 above)
Comment (Hurleveur) [reply source/Engine.cpp:1276]: Oh, dislexya striking again!

File: source/Government.cpp (hunk 3 above)
Comment (Hurleveur) [reply source/Government.cpp:263]: It shouldnt? Mb I got events wrong
You can just define them all under a node of raid if you want two of them

File: source/Panel.cpp (hunk 4 above)
Comment (Hurleveur) [reply source/Panel.cpp:26]: I had to, the style checker was screaming at me

Comment (quyykk) [reply source/Panel.cpp:26]: weird, let's blame @tibetiroka

File: source/Government.cpp (hunk 3 above)
Comment (quyykk) [reply source/Government.cpp:263]: oh yeah I see what you mean

File: source/Engine.cpp [hunk 5]
```diff
@@ -1261,20 +1261,27 @@ void Engine::EnterSystem()
...
+	{
+		const Government *raidGovernment = raidFleet.first->GetGovernment();
+		if(raidGovernment && raidGovernment->IsEnemy())
+		{
+			pair<double, double> factors = player.RaidFleetFactors();
+			double attraction = .005 * (factors.first - factors.second - raidFleet.second);
+			// Consider the other fleets in the local system.
+			// Do not take into account the raid fleet, that would just mess with the minimum attraction of it.
+			for(const auto &fleet : system->Fleets())
+				attraction += sqrt(fleet.Get()->Strength() / 10000.) / fleet.Period()
```
Comment (Hurleveur) [inline source/Engine.cpp:1274]:
```suggestion
				attraction -= sqrt(fleet.Get()->Strength() / 10000.) / fleet.Period()
```

**Comment by quyykk** [comment]
I think this should probably follow the syntax as others:

```
gov "gjk"
    raid <fleet> <#attraction>
    raid <fleet> <#attraction>
    # ...
```

?

**Comment by Hurleveur** [comment]
> I think this should probably follow the syntax as others:
> 
> ```
> gov "gjk"
>     raid <fleet> <#attraction>
>     raid <fleet> <#attraction>
>     # ...
> ```
> 
> ?

Done.

File: source/Government.h [hunk 6]
```diff
@@ -20,6 +20,7 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 #include "ExclusiveItem.h"
 #include "LocationFilter.h"
 
+#include <functional>
```
Comment (quyykk) [inline source/Government.h:23]: unneeded include

File: source/Government.cpp [hunk 7]
```diff
@@ -258,7 +263,21 @@ void Government::Load(const DataNode &node)
 		else if(key == "language")
 			language = child.Token(valueIndex);
 		else if(key == "raid")
-			raidFleet = GameData::Fleets().Get(child.Token(valueIndex));
+		{
+			const Fleet *raidingFleet = GameData::Fleets().Get(child.Token(valueIndex));
```
Comment (quyykk) [inline source/Government.cpp:267]: If in the second definition of this government a raid is defined, it should overwrite the previous raid definitions, for backwards compatibility. Unless `add` is used ofc.

File: source/Government.h (hunk 6 above)
Comment (Hurleveur) [reply source/Government.h:23]: its for std::pair

File: source/Government.cpp (hunk 7 above)
Comment (Hurleveur) [reply source/Government.cpp:267]: it does, above?
if(raidnotcleared && !add && !remove) does that I think

File: source/Government.h (hunk 6 above)
Comment (warp-core) [reply source/Government.h:23]: `std::pair` comes from the header `<utility>`.
`<functional>`: https://en.cppreference.com/w/cpp/header/functional

File: source/Government.cpp (hunk 7 above)
Comment (quyykk) [reply source/Government.cpp:267]: No, because that if above is guarded by if(remove). I'm talking in the scenario without any removes.

Comment (Hurleveur) [reply source/Government.cpp:272]: oh yeah bruh what I had written was nonsense xD

File: source/Government.h [hunk 8]
```diff
@@ -152,7 +154,8 @@ class Government {
 	const Phrase *hostileHail = nullptr;
 	const Phrase *hostileDisabledHail = nullptr;
 	std::string language;
-	const Fleet *raidFleet = nullptr;
+	bool notClearedRaids = true;
```
Comment (quyykk) [inline source/Government.h:157]: This can be a local variable

File: source/Government.h [hunk 9]
```diff
@@ -88,9 +89,10 @@ class Government {
 	std::string GetHail(bool isDisabled) const;
 	// Find out if this government speaks a different language.
 	const std::string &Language() const;
-	// Pirate raids in this government's systems use this fleet definition. If
-	// it is null, there are no pirate raids.
-	const Fleet *RaidFleet() const;
+	// Pirate raids in this government's systems use these fleet definitions. If
+	// it is empty, there are no pirate raids.
+	// The second attribute denotes the minimal attraction required to make the fleet come.
```
Comment (Amazinite) [inline source/Government.h:94]:
```suggestion
	// The second attribute denotes the minimal attraction required for the fleet to appear.
```

File: source/Government.cpp [hunk 10]
```diff
@@ -439,9 +463,9 @@ const string &Government::Language() const
 
 // Pirate raids in this government's systems use this fleet definition. If
 // it is null, there are no pirate raids.
```
Comment (Amazinite) [inline source/Government.cpp:465]:
```suggestion
// Pirate raids in this government's systems use these fleet definitions. If
// it is empty, there are no pirate raids.
// The second attribute denotes the minimal attraction required for the fleet to appear.
```

File: source/Government.cpp [hunk 11]
```diff
@@ -121,8 +123,11 @@ void Government::Load(const DataNode &node)
 		{
 			if(key == "provoked on scan")
 				provokedOnScan = false;
-			else if(key == "raid")
-				raidFleet = nullptr;
+			else if(key == "raid" && !hasValue)
+			{
+				raidFleets.clear();
+				notClearedRaids = false;
+			}
```
Comment (Amazinite) [inline source/Government.cpp:130]: The bottom of this if/else tree has a print trace saying you can't remove from the given value. We enter this tree if the line starts with `remove`, but we're skipping over this raid clear if we're given a specific raid to remove. That means if we have `remove raid "Pirate Raid"`, it'll enter `if(remove)`, skip over this `key == "raid"` because `!hasValue` is false, then print a trace saying you can't remove the given value, even if you can. Then, because the `if(remove)` is attached to all the other key checks by `else if`s, we never hit the normal `key == "raid"` code.

This is all set up this way because Government hasn't had a key that you can remove a single value of before. It's always either been overwrite what was there or completely clear it out. Look to a class like System for a place where `remove` behavior is different depending on if a value is given or not.

File: source/Government.cpp [hunk 12]
```diff
@@ -258,7 +263,26 @@ void Government::Load(const DataNode &node)
 		else if(key == "language")
 			language = child.Token(valueIndex);
 		else if(key == "raid")
-			raidFleet = GameData::Fleets().Get(child.Token(valueIndex));
+		{
+			if(notClearedRaids && !add && !remove)
```
Comment (Amazinite) [inline source/Government.cpp:267]: There's code above this point that prevents `add` from ever being true by the time we reach here:
```
else if(add)
	child.PrintTrace("Error: Unsupported use of add:");
```

File: source/Engine.cpp [hunk 13]
```diff
@@ -1261,20 +1261,27 @@ void Engine::EnterSystem()
...
+		const Government *raidGovernment = raidFleet.first->GetGovernment();
+		if(raidGovernment && raidGovernment->IsEnemy())
+		{
+			pair<double, double> factors = player.RaidFleetFactors();
+			double attraction = .005 * (factors.first - factors.second - raidFleet.second);
+			// Consider the other fleets in the local system.
+			// Do not take into account the raid fleet, that would just mess with the minimum attraction of it.
+			for(const auto &fleet : system->Fleets())
+				attraction -= sqrt(fleet.Get()->Strength() / 10000.) / fleet.Period()
+					* (fleet.Get()->GetGovernment()->IsEnemy(raidGovernment) ? 1. : -1.);
```
Comment (Amazinite) [inline source/Engine.cpp:1275]: Do you have any numbers on why this is the way you've done it? How much less likely is it to be attacked in Sol? How much more likely is it to be attacked in Shaula?

Why a multiplier of -1 for fleets that aren't enemies instead of 0? By using -1, fleets that are neutral to the raid fleet but might also be neutral to the player increase the change of getting jumped.

File: source/Government.cpp [hunk 14]
```diff
@@ -103,6 +103,8 @@ void Government::Load(const DataNode &node)
 			displayName = name;
 	}
 
+	bool notClearedRaids = true;
```
Comment (Amazinite) [inline source/Government.cpp:106]: No need for a double negative. Looking elsewhere, the variable would look like this instead:
```suggestion
	bool clearRaids = !raidFleets.isEmpty();
```

**Review by Amazinite — CHANGES_REQUESTED** [review]

File: source/Engine.cpp (hunk 13 above)
Comment (Hurleveur) [reply source/Engine.cpp:1275]: Since this is based on the spawnrate and not the actual fleets currently there so this can easily be calculated, I didnt do it though
About the 1 0 -1 I'll make it elegant

File: source/Government.cpp (hunk 11 above)
Comment (Hurleveur) [reply source/Government.cpp:130]: Cant remove in general have !hasValue added to it?

File: source/Government.cpp [hunk 15]
```diff
@@ -103,6 +103,9 @@ void Government::Load(const DataNode &node)
 			displayName = name;
 	}
 
+	// No need to consider clearing it if it's empty.
+	bool clearedRaids = !raidFleets.isEmpty();
```
Comment (Amazinite) [inline source/Government.cpp:107]: I meant what I said.
```suggestion
	// If raid fleets already exist, clear all existing raid fleets if the "add" or "remove" keywords aren't used
	// and a "raid" key is encountered.
	bool clearRaids = !raidFleets.isEmpty();
```

File: source/Government.cpp [hunk 16]
```diff
@@ -231,6 +234,27 @@ void Government::Load(const DataNode &node)
 			bribe = add ? bribe + child.Value(valueIndex) : child.Value(valueIndex);
 		else if(key == "fine")
 			fine = add ? fine + child.Value(valueIndex) : child.Value(valueIndex);
+		else if(key == "raid")
+		{
+			if(!clearedRaids && !add && !remove)
```
Comment (Amazinite) [inline source/Government.cpp:239]:
```suggestion
			if(clearRaids && !add && !remove)
```

File: source/Government.cpp [hunk 17]
```diff
@@ -231,6 +234,27 @@ void Government::Load(const DataNode &node)
...
+			{
+				raidFleets.clear();
+				clearedRaids = true;
```
Comment (Amazinite) [inline source/Government.cpp:242]:
```suggestion
				clearRaids = false;
```

File: source/Government.cpp (hunk 11 above)
Comment (Amazinite) [reply source/Government.cpp]: That should work, yes.

File: source/Engine.cpp (hunk 13 above)
Comment (Hurleveur) [reply source/Engine.cpp]: Only question that remains is how much it should matter, rn I made it `(raidStrength / (fleet.Get()->Strength() / fleet.Period()))` so it wont matter too much but mb we should multiply the other fleet by smth
could always be tuned in a later PR, and eventually be made part of the gamerules

File: source/Government.cpp (hunk 11 above)
Comment (Hurleveur) [reply source/Government.cpp]: Actually that would make the error parsing less precise than it used to be I think.
So I made it the same way as it works in system +- (no need for a whole vector here)

File: source/Government.cpp (hunk 15 above)
Comment (Amazinite) [reply source/Government.cpp]: 😔

File: source/Engine.cpp (hunk 13 above)
Comment (Amazinite) [reply source/Engine.cpp]: Flip it to `fleet strength / raid strength`. A stronger system fleet should *increase* the multiplier on how hard attraction is hit, not decrease it.

File: source/PlayerInfo.cpp [hunk 18]
```diff
@@ -2925,6 +2959,22 @@ void PlayerInfo::RegisterDerivedConditions()
...
+			return 0;
+
+		double totalAttraction = 0;
+		for(const auto &raidFleet : system->GetGovernment()->RaidFleets())
+			totalAttraction += RaidFleetAttraction(raidFleet, system);
+		
+		return round(totalAttraction * 1000.);
+	};
+	systemAttractionProvider.SetGetFunction(systemAttractionFun);
+	systemAttractionProvider.SetHasFunction(systemAttractionFun);
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:2976]: I've realized that this doesn't really work with multiple fleets. Consider the following scenario:
* System X has a single raid fleet while system Y has two raid fleets.
* In X, the fleet's attraction toward the player is 30. In Y, each fleet has an attraction of 15.
* The actual likelihood of any single fleet attacking you is `p(x) = 1 - (1 - x/1000)^10` (in English, one minus the probability for 10 of 10 attempts to spawn this fleet to fail) where x is the attraction value from this condition, as x is actually the odds of any one instance of the raid fleet attacking, and up to 10 raid fleets of a single type can be spawned at once.
* For system X, p(30) = 26.26%.
* For system Y, it's a little more complicated. We can't just take 2 * p(15), as that's not how these probabilistic events add together. What we actually have to do is take `1 - (1 - p(15)) * (1 - p(15))`, or `1 - (1 - p(15))^2` (in English, one minus the odds that the first fleet doesn't attack times the odds that the second fleet doesn't attack), which comes out to 26.09%. 
* So while both of these systems would return the same attraction value, they don't have the same chance of encountering a raid.

As such, assuming I haven't messed up my probabilities (please double check me), I think this needs to just return the raid chance as follows.
```suggestion
	auto &&systemAttractionProvider = conditions.GetProviderPrefixed("raid chance in system: ");
	auto systemAttractionFun = [this](const string &name) -> double
	{
		const System *system = GameData::Systems().Find(name.substr(strlen("raid chance in system: ")));
		if(!system)
			return 0.;

		// This variable represents the probability of no raid fleets spawning.
		double chance = 1.;
		for(const auto &raidFleet : system->GetGovernment()->RaidFleets())
		{
			// The attraction is the % chance for a single instance of this fleet to appear.
			double attraction = RaidFleetAttraction(raidFleet, system);
			// Calculate the % chance for no instances to appear from 10 rolls.
			double prob = pow(1. - attraction, 10.);
			// The chance of neither of two fleets appearing is the chance of the first not appearing
			// times the chance of the second not appearing.
			chance *= prob;
		}
		// The probability of any single fleet appearing is 1 - chance.
		return round((1. - chance) * 1000.);
	};
	systemAttractionProvider.SetGetFunction(systemAttractionFun);
	systemAttractionProvider.SetHasFunction(systemAttractionFun);
```

File: source/PlayerInfo.cpp [hunk 19]
```diff
@@ -1225,6 +1225,40 @@ pair<double, double> PlayerInfo::RaidFleetFactors() const
...
+	const Government *raidGov = raidFleet.first->GetGovernment();
+	if(raidGov && raidGov->IsEnemy())
+	{
+		// The player's base attraction to a fleet is determined by their fleet attraction minus
+		// their fleet deterence, minus whatever the minimum attraction of this raid fleet is.
+		pair<double, double> factors = RaidFleetFactors();
+		attraction = .005 * (factors.first - factors.second - raidFleet.second);
+		// Then we consider the strength of other fleets in the system.
+		auto raidStrength = raidFleet.first->Strength();
+		if(system)
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:1240]: Just to be safe.
```suggestion
		if(system && raidStrength)
```

File: source/PlayerInfo.cpp [hunk 20]
```diff
@@ -1225,6 +1225,40 @@ pair<double, double> PlayerInfo::RaidFleetFactors() const
...
+					// raid attraction will increase. If the fleet is hostile to only the raid fleet,
+					// the raid attraction will decrease. The amount of increase or decrease is determined
+					// by the strength of the fleet relative to the raid fleet. System fleets which are
+					// stronger have a larger impact.
+					double strength = fleet.Get()->Strength() / fleet.Period();
+					attraction -= (gov->IsEnemy(raidGov) - gov->IsEnemy()) * (strength / raidStrength);
+				}
+			}
+	}
+	return min(0., attraction);
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:1257]: Meant to do this.
```suggestion
	return max(0., min(1., attraction));
```

**Comment by Amazinite** [comment]
Should we have a range of attractions that fleets spawn within? That is, allow a max attraction that prevents fleets from spawning if you're above a certain point. This way we could more easily have a ramping of the type of fleet that attacks you without having a bunch of small fleets spawn when one large fleet spawns.
```
raid <fleet> [<min attraction> [<max attraction>]]
```

File: source/PlayerInfo.cpp (hunk 18 above)
Comment (Hurleveur) [reply source/PlayerInfo.cpp:2976]: Oh I didnt understand you meant to return the probability of attack itself, makes sense though

**Comment by Hurleveur** [comment]
Makes sense, I'll add that too
Also will look a bit more at government caus I dont think it makes sense to look for "objects"

File: source/PlayerInfo.cpp (hunk 18 above)
Comment (Hurleveur) [reply source/PlayerInfo.cpp:2976]: Shouldnt pow check for (100-attraction) given we check attraction against a real that is between 0 and 100 (I think, right?)

**Comment by Amazinite** [comment]
> Also will look a bit more at government caus I dont think it makes sense to look for "objects"

Oops. That was a copy/paste from System.

**Comment by Hurleveur** [comment]
Good that I removed it then.
Should be good to go once this passes the tests.

File: source/Government.h [hunk 21]
```diff
@@ -152,7 +154,7 @@ class Government {
 	const Phrase *hostileHail = nullptr;
 	const Phrase *hostileDisabledHail = nullptr;
 	std::string language;
-	const Fleet *raidFleet = nullptr;
+	std::vector<std::pair<const Fleet *, std::pair<double, double>>> raidFleets;
```
Comment (Amazinite) [inline source/Government.h:157]: It'd probably be cleaner at this point for Government to just have an inner class.
```
std::vector<RaidFleet> raidFleets;

class RaidFleet {
public:
	RaidFleet(const Fleet *fleet, double minAttraction, double maxAttraction);
	const Fleet *GetFleet() const;
	double MinAttraction() const;
	double MaxAttraction() const;

private:
	const Fleet *fleet = nullptr;
	double minAttraction;
	double maxAttraction;
}
```

**Comment by Amazinite** [comment]
> Good that I removed it then.
> Should be good to go once this passes the tests.

I'll give it a final look later today if I'm not dead from staying up until 3am all week.

File: source/Government.h (hunk 21 above)
Comment (Hurleveur) [reply source/Government.h:157]: ~~I suspected you'd ask this~~
Alr then

**Comment by Hurleveur** [comment]
Oh wow you should get some rest this doesnt matter

File: source/PlayerInfo.cpp [hunk 22]
```diff
@@ -2925,6 +2963,31 @@ void PlayerInfo::RegisterDerivedConditions()
...
+			// The attraction is the % chance for a single instance of this fleet to appear.
+			double attraction = RaidFleetAttraction(raidFleet, system);
+			// Calculate the % chance for no instances to appear from 10 rolls.
+			double noFleetProb = pow(1. - attraction, 10.);
+			// The chance of neither of two fleets appearing is the chance of the first not appearing
+			// times the chance of the second not appearing.
+			safeChance *= noFleetProb;
+		}
+		// The probability of any single fleet appearing is 1 - chance.
+		return round((1. - safeChance) * 1000.);
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:2986]: this is what derpy was speaking of checking @warp-core

File: source/Government.cpp [hunk 23]
```diff
@@ -148,8 +188,34 @@ void Government::Load(const DataNode &node)
...
+				for(auto it = raidFleets.begin(); it != raidFleets.end(); )
+					if(it->GetFleet() == fleet)
+						it = raidFleets.erase(it);
+					else
+						++it;
+			}
+			else
+				raidFleets.emplace_back(RaidFleet(fleet,
+					(child.Size() > (valueIndex + 1) ? child.Value(valueIndex + 1) : 2.),
+					child.Size() > (valueIndex + 2) ? child.Value(valueIndex + 2) : 0.));
```
Comment (Amazinite) [inline source/Government.cpp:214]: You don't need to construct before emplacing into a vector. `emplace_back` does that for you.
```suggestion
				raidFleets.emplace_back(fleet,
					child.Size() > (valueIndex + 1) ? child.Value(valueIndex + 1) : 2.,
					child.Size() > (valueIndex + 2) ? child.Value(valueIndex + 2) : 0.);
```

File: source/PlayerInfo.cpp [hunk 24]
```diff
@@ -1225,6 +1225,44 @@ pair<double, double> PlayerInfo::RaidFleetFactors() const
 
 
 
+double PlayerInfo::RaidFleetAttraction(const Government::RaidFleet &raidFleet, const System *system) const
+{
+	double attraction = 0.;
+	const Government *raidGov = raidFleet.GetFleet()->GetGovernment();
+	if(raidGov && raidGov->IsEnemy())
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:1232]:
```suggestion
double PlayerInfo::RaidFleetAttraction(const Government::RaidFleet &raid, const System *system) const
{
	double attraction = 0.;
	const Fleet *raidFleet = raid.GetFleet();
	const Government *raidGov = raidFleet ? raidFleet->GetGovernment() : nullptr;
	if(raidGov && raidGov->IsEnemy())
```

File: source/PlayerInfo.cpp [hunk 25]
```diff
@@ -1225,6 +1225,44 @@ pair<double, double> PlayerInfo::RaidFleetFactors() const
...
+	{
+		// The player's base attraction to a fleet is determined by their fleet attraction minus
+		// their fleet deterence, minus whatever the minimum attraction of this raid fleet is.
+		pair<double, double> factors = RaidFleetFactors();
+		// If there is a maximum attraction for this fleet, and we are above it, it will not spawn.
+		if(raidFleet.MaxAttraction() && factors.first > raidFleet.MaxAttraction())
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:1238]:
```suggestion
		if(raid.MaxAttraction() && factors.first > raid.MaxAttraction())
```

File: source/PlayerInfo.cpp [hunk 26]
```diff
@@ -1225,6 +1225,44 @@ pair<double, double> PlayerInfo::RaidFleetFactors() const
...
+			return 0;
+
+		attraction = .005 * (factors.first - factors.second - raidFleet.MinAttraction());
+		// Then we consider the strength of other fleets in the system.
+		auto raidStrength = raidFleet.GetFleet()->Strength();
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:1243]:
```suggestion
		int64_t raidStrength = raidFleet->Strength();
```

File: source/PlayerInfo.cpp [hunk 27]
```diff
@@ -1225,6 +1225,45 @@ pair<double, double> PlayerInfo::RaidFleetFactors() const
...
+	double attraction = 0.;
+	const Fleet *raidFleet = raid.GetFleet();
+	const Government *raidGov = raidFleet ? raidFleet->GetGovernment() : nullptr;
+	if(raidGov && raidGov->IsEnemy())
+	{
+		// The player's base attraction to a fleet is determined by their fleet attraction minus
+		// their fleet deterence, minus whatever the minimum attraction of this raid fleet is.
+		pair<double, double> factors = RaidFleetFactors();
+		// If there is a maximum attraction for this fleet, and we are above it, it will not spawn.
+		if(raid.MaxAttraction() && factors.first > raid.MaxAttraction())
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:1239]:
```suggestion
		if(raid.MaxAttraction() > 0 && factors.first > raid.MaxAttraction())
```

File: source/PlayerInfo.cpp [hunk 28]
```diff
@@ -1225,6 +1225,45 @@ pair<double, double> PlayerInfo::RaidFleetFactors() const
...
+	if(raidGov && raidGov->IsEnemy())
+	{
+		// The player's base attraction to a fleet is determined by their fleet attraction minus
+		// their fleet deterence, minus whatever the minimum attraction of this raid fleet is.
+		pair<double, double> factors = RaidFleetFactors();
+		// If there is a maximum attraction for this fleet, and we are above it, it will not spawn.
+		if(raid.MaxAttraction() > 0 && factors.first > raid.MaxAttraction())
+			return 0;
+
+		attraction = .005 * (factors.first - factors.second - raidFleet->MinAttraction());
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:1242]:
```suggestion
		attraction = .005 * (factors.first - factors.second - raid->MinAttraction());
```

**Review by Amazinite — APPROVED** [review]
